pkg/util/time_cache_test.go

```diff
@@ -115,3 +117,67 @@ func TestCacheParallelOneCall(t *testing.T) {
 		t.Errorf("Expected %v, got %v", e, a)
 	}
 }
+
+func TestCacheParallelNoDeadlocksNoDoubleCalls(t *testing.T) {
+	// Make 50 random keys
+	keys := []string{}
+	fuzz.New().NilChance(0).NumElements(50, 50).Fuzz(&keys)
+
+	// Data structure for tracking when each key is accessed.
+	type callTrack struct {
+		sync.Mutex
+		accessTimes []time.Time
+	}
+	calls := map[string]*callTrack{}
+	for _, k := range keys {
+		calls[k] = &callTrack{}
+	}
+
+	// This is called to fill the cache in the case of a cache miss
+	// or cache entry expiration. We record the time.
+	ff := func(key string) T {
+		ct := calls[key]
+		ct.Lock()
+		ct.accessTimes = append(ct.accessTimes, time.Now())
+		ct.Unlock()
+		// make sure that there is time for multiple requests to come in
+		// for the same key before this returns.
+		time.Sleep(time.Millisecond)
+		return key
+	}
+
+	cacheDur := 10 * time.Millisecond
+	c := NewTimeCache(RealClock{}, cacheDur, ff)
+
+	// Spawn a bunch of goroutines, each of which sequentially requests
+	// 500 random keys from the cache.
+	runtime.GOMAXPROCS(16)
+	var wg sync.WaitGroup
+	for i := 0; i < 500; i++ {
+		wg.Add(1)
+		go func(seed int64) {
+			r := rand.New(rand.NewSource(seed))
+			for i := 0; i < 500; i++ {
+				c.Get(keys[r.Intn(len(keys))])
+			}
+			wg.Done()
+		}(rand.Int63())
+	}
+	wg.Wait()
+
+	// Since the cache should hold things for 10ms, no calls for a given key
+	// should be more closely spaced than that.
+	for k, ct := range calls {
+		if len(ct.accessTimes) < 2 {
+			continue
+		}
+		cur := ct.accessTimes[0]
+		for i := 1; i < len(ct.accessTimes); i++ {
+			next := ct.accessTimes[i]
+			if next.Sub(cur) < cacheDur {
```

Comment:
Does this really work reliably at this timescale? What about NTP setting back the clock, or TSC changes due to rescheduling to a different socket? My experience is that one can't count on clock monotonicity except at the granularity of O(seconds).

I'm willing to merge this and wait and see, but I wouldn't be surprised if it proved to be flaky.